src/transformers/trainer_utils.py

```diff
@@ -1,6 +1,77 @@
+import logging
+import os
 from typing import Dict, NamedTuple, Optional
 
 import numpy as np
+from tqdm.auto import tqdm
+
+
+logger = logging.getLogger(__name__)
+
+try:
+    import wandb
+
+    wandb.ensure_configured()
+    if wandb.api.api_key is None:
+        _has_wandb = False
+        wandb.termwarn("W&B installed but not logged in.  Run `wandb login` or set the WANDB_API_KEY env variable.")
+    else:
+        _has_wandb = False if os.getenv("WANDB_DISABLED") else True
+except (ImportError, AttributeError):
+    _has_wandb = False
+    logger.info(
+        "You are instantiating a Trainer but W&B is not installed. To use wandb logging, "
+        "run `pip install wandb; wandb login` see https://docs.wandb.com/huggingface."
+    )
+
+
+def is_wandb_available():
+    return _has_wandb
+
+
+def setup_wandb(trainer):
+    """
+    Setup the optional Weights & Biases (`wandb`) integration.
+
+    One can override this method to customize the setup if needed.  Find more information at https://docs.wandb.com/huggingface
+    You can also override the following environment variables:
+
+    Environment:
+        WANDB_WATCH:
+            (Optional, ["gradients", "all", "false"]) "gradients" by default, set to "false" to disable gradient logging
+            or "all" to log gradients and parameters
+        WANDB_PROJECT:
+            (Optional): str - "huggingface" by default, set this to a custom string to store results in a different project
+        WANDB_DISABLED:
+            (Optional): boolean - defaults to false, set to "true" to disable wandb entirely
+    """
+    logger.info('Automatic Weights & Biases logging enabled, to disable set os.environ["WANDB_DISABLED"] = "true"')
+
+    wandb.init(project=os.getenv("WANDB_PROJECT", "huggingface"), config=vars(trainer.args))
+    # keep track of model topology and gradients (only with Pytorch)
+    if os.getenv("WANDB_WATCH") != "false" and trainer.__class__.__name__ == "Trainer":
+        wandb.watch(
+            trainer.model, log=os.getenv("WANDB_WATCH", "gradients"), log_freq=max(100, trainer.args.logging_steps)
+        )
+    # give access to wandb module
+    trainer._wandb = wandb
+
+
+def log_metrics(trainer, logs: Dict[str, float], iterator: Optional[tqdm] = None) -> None:
```

Comment:
Not sure we want to share that accross both Trainer to be honest, this seems weird

Reply:
Are you talking about both the setup and logging part?

Initially I didn't notice tensorboard logging would be different between the 2 methods but now it may make sense to keep entire logging separate.

What about setting up wandb (I would just add a bool "watch" argument to make it clearer).